src/policy/fees.cpp

```diff
@@ -652,6 +652,23 @@ void CBlockPolicyEstimator::processBlock(unsigned int nBlockHeight,
     shortStats->UpdateMovingAverages();
     longStats->UpdateMovingAverages();
 
+    // Do not consider transactions that were CPFP'd.
+    // For each mempool transaction that was included in a block, if its individual feerate is higher
+    // than its ancestor score it incentivized the inclusion of some of its parents in the block.
+    // Remove all such parents whose feerate is lower than the child's of the tracking before updating
+    // the data points.
+    for (const auto& entry : entries) {
+        CFeeRate individual_feerate{entry->GetFee(), (uint32_t)entry->GetTxSize()};
+        CFeeRate ancestor_score{entry->GetModFeesWithAncestors(), (uint32_t)entry->GetSizeWithAncestors()};
+
+        if (individual_feerate > ancestor_score) {
```

Comment:
Only having `parent_feerate < individual_feerate` does not account for the size of the dependencies. For instance, a transaction `C` (2000 sat, 100 vb) spending a transaction `B` (200 sat, 1000 vb) which in turn spends a transaction `A` (1000sat, 100 vb):
```
A --- B --- C
```
`C` clearly does not CPFP `A`, while it has a higher feerate. Thanks for bringing this up, the current logic is flawed in this sense. We should be checking `parent_feerate < ancestor_score`!

Which explains the check that the `individual_feerate > ancestor_score`. It is to make sure the child is paying, to prevent a pay-for-sibling scenario:
```
A
  \
     C
  /
B
```
`B` is 2sat/vb, `A` is 1sat/vb, `C` is 1sat/vb. The ancestor score of `C` is larger than the feerate of `A`, but it's not a CPFP.

So this may be too conservative in some scenario, such as the one you describe. But i can't think of a reasonable way to take it into account. Also i think we better be overly conservative if that gives us confidence we won't let some false positives through.

Reply:
Resolving this, i don't think we can get rid of this heuristic. https://github.com/bitcoin/bitcoin/pull/25380#issuecomment-1494455306 gives a counter example where it would lead to a false positive. Please let me know if you disagree.